src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +574,226 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
 	if (HeapTupleIsValid(namespaceTuple))
 	{
 		ReleaseSysCache(namespaceTuple);
-
 		ereport(ERROR, (errmsg("schema is expected to be already dropped "
 							   "because this function is only expected to "
 							   "be called from Citus drop hook")));
 	}
-
 	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);
 
 	DeleteTenantSchemaLocally(schemaId);
 	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));
 
 	DeleteColocationGroup(tenantSchemaColocationId);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_distribute gets a regular schema name, then converts it to a tenant
+ * schema.
+ */
+Datum
+citus_schema_distribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Ensure schema checks */
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
```

Comment:
should we check this before locking the schema?

Reply:
We should check after the lock anyway to verify the schema is not altered or dropped. After the lock we directly check and throw error if it is not the schema owner who performed the operation.

Reply:
But wouldn't this allow malicious users to lock the schema unnecessarily? I believe this is also what we do in various functions (such as create_distributed_table?), @marcocitus can correct me if I'm not remembering this correctly.

Reply:
(If we decide moving this check to the line before LockDatabaseObject, we should do the same for other udf too)

Reply:
Actually yes, they may call this udf inside a transaction block and acquire the lock. Then we should also check at the start.

Reply:
Did you have a chance to check with @marcocitus that we're okay with locking the schema before checking the owner?